range_queries/heavy_light_decomposition.cpp

```diff
@@ -0,0 +1,579 @@
+	*/
+	void set_node_val(std::vector<X> node_val) {
+		assert(static_cast<int>(node_val.size()) == t_nodes);
+		t_val = node_val;
+	}
+
+	/**
+	 * @brief This function must be called after the tree adjacency list and node values are populated
+	 * The function initializes the required parametes, and populates the segment tree
+	*/
+	void init() {
+		assert(t_nodes > 0);
+		dfs_size(t_root);
+		dfs_lca(t_root);
+	}
+
+	/**
+	 * @brief The function lifts a node, k units up the tree.
+	 * The lifting is done in place, and the result is stored in the address pointed by p.
+	 * @param p a pointer to the variable that stores the node id
+	 * @param dist the distance to move up the tree
+	 * @return void.
+	*/
+	void lift(int* const p, int dist) {
+		for(int k = 0; k < t_maxlift; k++) {
+			if(*p == -1) return;
+			if(dist & 1) {
+				*p = t_par[*p][k];
+			}
+			dist >>= 1;
+		}
+	} 
+	
+	/**
+	 * @brief The function returns the kth ancestor of a node
+	 * @param p the node id whose kth ancestor is to be found
+	 * @param dist the distance to move up the tree
+	 * @return the kth ancestor of node
+	*/
+	int kth_ancestor(int p, const int& dist) {
+		lift(&p, dist);
+		return p;
+	}
+
+	/**
+	 * @brief The function returns the least common ancestor of two nodes
+	 * @param a node id_1
+	 * @param b node id_2
+	 * @return the least common ancestor of node a, and node b
+	*/
+	int lca(int a, int b) {
+		assert(a >= 0 and b >= 0 and a < t_nodes and b < t_nodes);
+		if(t_depth[a] > t_depth[b]) {
+			lift(&a, t_depth[a] - t_depth[b]);
+		}
+		if(t_depth[b] > t_depth[a]) {
+			lift(&b, t_depth[b] - t_depth[a]);
+		}
+		if(a == b) return a;
+		for(int k = 0; k < t_maxlift; k++) {
+			if(t_par[a][k] != t_par[b][k]) {
+				a = t_par[a][k];
+				b = t_par[b][k];
+			}
+		}
+		return t_par[a][0];
+	}
+};
+
+/** 
+ * @brief Segment Tree, to store heavy chains
+ * @tparam the data type of the values stored in the tree nodes
+*/
+template<typename X>
+class SG {
+private:
+	/**
+	 * @brief Everything here is private,
+	 * and can only be acced thought the methods,
+	 * in the derived class (HLD).
+	*/
+
+	std::vector<X> s_tree;
+	int s_size;
+	template<typename T> friend class HLD;
+	X sret_init = 0;
+	
+	/**
+	 * @brief Function that specifies the type of operation involved when segments are combined
+	 * @param lhs the left segment
+	 * @param rhs the right segment
+	 * @return the combined result
+	*/
+	X combine(X lhs, X rhs) {
+		return lhs + rhs;
+	}
+
+
+	/**
+	 * @brief Class parameterized constructor. Resizes the and initilizes the data members.
+	 * @param nodes the total number of nodes in the tree
+	*/
+	explicit SG(int size) {
+		s_size = size;
+		s_tree.assign(2*s_size, 0ll);
+	}
+
+	/**
+	 * @brief Update the value at a node
+	 * @param p the node to be udpated
+	 * @param v the update value
+	*/
+	void update(int p, X v) {
+		for(p += s_size; p > 0; p >>= 1) s_tree[p] += v;
```

Comment:
Make it look like C++ code, not C code.
```suggestion
		for(p += s_size; p > 0; p >>= 1) { s_tree[p] += v; }
```